File: frontend/src/services/contract.ts
import apiClient from '@/api/client';

export interface ContractTemplate {
  id: string;
  code: string;
  name: string;
  description?: string;
  contract_type: string;
  category: string;
  content: string;
  variables?: string;
  version: string;
  is_active: boolean;
  is_system: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface CreateContractTemplate {
  code: string;
  name: string;
  description?: string;
  contract_type: string;
  category: string;
  content: string;
  variables?: string;
  version?: string;
  is_active?: boolean;
  is_system?: boolean;
  sort_order?: number;
}

export interface UpdateContractTemplate {
  name?: string;
  description?: string;
  contract_type?: string;
  category?: string;
  content?: string;
  variables?: string;
  version?: string;
  is_active?: boolean;
  sort_order?: number;
}

export interface Contract {
  id: string;
  code: string;
  name: string;
  description?: string;
  contract_type: string;
  category: string;
  content?: string;
  content_rendered?: string;
  status: string;
  first_party: string;
  second_party: string;
  amount?: number;
  currency: string;
  start_date?: string;
  end_date?: string;
  sign_date?: string;
  project_id?: string;
  created_by?: string;
  risk_level?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateContract {
  code?: string;
  name: string;
  description?: string;
  contract_type: string;
  category: string;
  content?: string;
  template_id?: string;
  content_rendered?: string;
  first_party: string;
  second_party: string;
  amount?: number;
  currency?: string;
  start_date?: string;
  end_date?: string;
  sign_date?: string;
  project_id?: string;
  created_by?: string;
  risk_level?: string;
  status?: string;
}

export interface UpdateContract {
  name?: string;
  description?: string;
  contract_type?: string;
  category?: string;
  content?: string;
  content_rendered?: string;
  first_party?: string;
  second_party?: string;
  amount?: number;
  currency?: string;
  start_date?: string;
  end_date?: string;
  sign_date?: string;
  project_id?: string;
  status?: string;
  risk_level?: string;
}

const API_BASE_URL = '/v1';

export const contractService = {
  async getContractTemplates(contractType?: string, category?: string, isActive?: boolean): Promise<ContractTemplate[]> {
    const params = new URLSearchParams();
    if (contractType) params.append('contract_type', contractType);
    if (category) params.append('category', category);
    if (isActive !== undefined) params.append('is_active', String(isActive));
    const response = await apiClient.get(`${API_BASE_URL}/contracts/templates?${params.toString()}`);
    return response.data?.data || response.data || [];
  },

  async getContractTemplate(id: string): Promise<ContractTemplate> {
    const response = await apiClient.get(`${API_BASE_URL}/contracts/templates/${id}`);
    return response.data?.data || response.data;
  },

  async createContractTemplate(data: CreateContractTemplate): Promise<ContractTemplate> {
    const response = await apiClient.post(`${API_BASE_URL}/contracts/templates`, data);
    return response.data?.data || response.data;
  },

  async updateContractTemplate(id: string, data: UpdateContractTemplate): Promise<ContractTemplate> {
    const response = await apiClient.patch(`${API_BASE_URL}/contracts/templates/${id}`, data);
    return response.data?.data || response.data;
  },

  async deleteContractTemplate(id: string): Promise<void> {
    await apiClient.delete(`${API_BASE_URL}/contracts/templates/${id}`);
  },

  async getContracts(status?: string, contractType?: string, category?: string, page?: number, pageSize?: number): Promise<Contract[]> {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    if (contractType) params.append('contract_type', contractType);
    if (category) params.append('category', category);
    if (page !== undefined) params.append('page', String(page));
    if (pageSize !== undefined) params.append('page_size', String(pageSize));
    const response = await apiClient.get(`${API_BASE_URL}/contracts?${params.toString()}`);
    return response.data?.data || response.data || [];
  },

  async getContract(id: string): Promise<Contract> {
    const response = await apiClient.get(`${API_BASE_URL}/contracts/${id}`);
    return response.data?.data || response.data;
  },

  async createContract(data: CreateContract): Promise<Contract> {
    const response = await apiClient.post(`${API_BASE_URL}/contracts`, data);
    return response.data?.data || response.data;
  },

  async updateContract(id: string, data: UpdateContract): Promise<Contract> {
    const response = await apiClient.patch(`${API_BASE_URL}/contracts/${id}`, data);
    return response.data?.data || response.data;
  },

  async deleteContract(id: string): Promise<void> {
    await apiClient.delete(`${API_BASE_URL}/contracts/${id}`);
  },
};

export const contractApi = contractService;

File: backend/core/src/db/contract.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ContractTemplate {
    pub id: Uuid,
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    pub contract_type: String,
    pub category: String,
    pub content: String,
    pub variables: Option<String>,
    pub version: String,
    pub is_active: bool,
    pub is_system: bool,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateContractTemplate {
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    pub contract_type: String,
    pub category: String,
    pub content: String,
    pub variables: Option<String>,
    pub version: Option<String>,
    pub is_active: Option<bool>,
    pub is_system: Option<bool>,
    pub sort_order: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateContractTemplate {
    pub name: Option<String>,
    pub description: Option<String>,
    pub contract_type: Option<String>,
    pub category: Option<String>,
    pub content: Option<String>,
    pub variables: Option<String>,
    pub version: Option<String>,
    pub is_active: Option<bool>,
    pub sort_order: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Contract {
    pub id: Uuid,
    pub code: String,
    pub name: String,
    pub description: Option<String>,
    pub contract_type: String,
    pub category: String,
    pub content: String,
    pub content_rendered: String,
    pub status: String,
    pub first_party: String,
    pub second_party: String,
    pub amount: Option<f64>,
    pub currency: String,
    pub start_date: Option<DateTime<Utc>>,
    pub end_date: Option<DateTime<Utc>>,
    pub sign_date: Option<DateTime<Utc>>,
    pub project_id: Option<Uuid>,
    pub created_by: Option<Uuid>,
    pub risk_level: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateContract {
    pub code: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub contract_type: String,
    pub category: String,
    pub content: Option<String>,
    pub template_id: Option<Uuid>,
    pub content_rendered: Option<String>,
    pub first_party: String,
    pub second_party: String,
    pub amount: Option<f64>,
    pub currency: Option<String>,
    pub start_date: Option<DateTime<Utc>>,
    pub end_date: Option<DateTime<Utc>>,
    pub sign_date: Option<DateTime<Utc>>,
    pub project_id: Option<Uuid>,
    pub created_by: Option<Uuid>,
    pub risk_level: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateContract {
    pub name: Option<String>,
    pub description: Option<String>,
    pub contract_type: Option<String>,
    pub category: Option<String>,
    pub content: Option<String>,
    pub content_rendered: Option<String>,
    pub first_party: Option<String>,
    pub second_party: Option<String>,
    pub amount: Option<f64>,
    pub currency: Option<String>,
    pub start_date: Option<DateTime<Utc>>,
    pub end_date: Option<DateTime<Utc>>,
    pub sign_date: Option<DateTime<Utc>>,
    pub project_id: Option<Uuid>,
    pub status: Option<String>,
    pub risk_level: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContractVariableValue {
    pub id: Uuid,
    pub contract_id: Uuid,
    pub variable_key: String,
    pub variable_value: String,
    pub variable_type: String,
    pub created_at: DateTime<Utc>,
}
